src/hotspot/share/opto/type.cpp

```diff
@@ -1782,264 +1693,125 @@ bool TypeInt::is_finite() const {
   return true;
 }
 
-//------------------------------dump2------------------------------------------
-// Dump TypeInt
-#ifndef PRODUCT
-static const char* intname(char* buf, size_t buf_size, jint n) {
-  if (n == min_jint)
-    return "min";
-  else if (n < min_jint + 10000)
-    os::snprintf_checked(buf, buf_size, "min+" INT32_FORMAT, n - min_jint);
-  else if (n == max_jint)
-    return "max";
-  else if (n > max_jint - 10000)
-    os::snprintf_checked(buf, buf_size, "max-" INT32_FORMAT, max_jint - n);
-  else
-    os::snprintf_checked(buf, buf_size, INT32_FORMAT, n);
-  return buf;
-}
-
-void TypeInt::dump2( Dict &d, uint depth, outputStream *st ) const {
-  char buf[40], buf2[40];
-  if (_lo == min_jint && _hi == max_jint)
-    st->print("int");
-  else if (is_con())
-    st->print("int:%s", intname(buf, sizeof(buf), get_con()));
-  else if (_lo == BOOL->_lo && _hi == BOOL->_hi)
-    st->print("bool");
-  else if (_lo == BYTE->_lo && _hi == BYTE->_hi)
-    st->print("byte");
-  else if (_lo == CHAR->_lo && _hi == CHAR->_hi)
-    st->print("char");
-  else if (_lo == SHORT->_lo && _hi == SHORT->_hi)
-    st->print("short");
-  else if (_hi == max_jint)
-    st->print("int:>=%s", intname(buf, sizeof(buf), _lo));
-  else if (_lo == min_jint)
-    st->print("int:<=%s", intname(buf, sizeof(buf), _hi));
-  else
-    st->print("int:%s..%s", intname(buf, sizeof(buf), _lo), intname(buf2, sizeof(buf2), _hi));
-
-  if (_widen != 0 && this != TypeInt::INT)
-    st->print(":%.*s", _widen, "wwww");
-}
-#endif
-
 //------------------------------singleton--------------------------------------
 // TRUE if Type is a singleton type, FALSE otherwise.   Singletons are simple
 // constants.
 bool TypeInt::singleton(void) const {
-  return _lo >= _hi;
+  return _lo == _hi;
 }
 
 bool TypeInt::empty(void) const {
-  return _lo > _hi;
+  return false;
 }
```

Comment:
I would like to see assert(_lo <= _hi) here since you unconditionally return `false`.

Reply:
`_lo <= _hi` (and other invariants) is checked in `verify_constraints` when creating the `TypeInt` instance.

Reply:
The call:

https://github.com/openjdk/jdk/blob/5990628a8337b9040128857f34359b169326eb23/src/hotspot/share/opto/type.cpp#L1598

And the implementation:

https://github.com/openjdk/jdk/blob/5990628a8337b9040128857f34359b169326eb23/src/hotspot/share/opto/rangeinference.cpp#L358

Notice that `this->contains(_srange._lo)` implies that `_lo <= _hi`

Reply:
Okay but assert will guard against future incorrect changes by someone who may not familiar with verification code you pointed. Some new path in type construction can be introduced which bypath those checks.

Reply:
My thought is that a `TypeInt` has many invariants and it would be really expensive if we check all of them at all use sites, and it seems not enough if we only check for one of those invariants. What do you think?